File: domain/src/main/java/by/rekuts/travelagency/domain/HotelEnumArrayType.java
package by.rekuts.travelagency.domain;

import org.hibernate.engine.spi.SharedSessionContractImplementor;
import org.hibernate.usertype.UserType;

import java.io.Serializable;
import java.sql.*;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class HotelEnumArrayType implements UserType {

    private static final int SQL_TYPE = Types.ARRAY;

    @Override
    public Object nullSafeGet(ResultSet rs, String[] strings, SharedSessionContractImplementor sharedSessionContractImplementor, Object o) throws SQLException {
        var dbArray = rs.getArray(strings[0]);
        var array = (Object[]) dbArray.getArray();
        var features = new ArrayList<>();
        for(Object ob : array) {
            for(var f : Hotel.Features.values()) {
                if (f.getValue().equals(ob.toString())) {
                    features.add(ob.toString());
                }
            }
        }
        return features;
    }

    @Override
    public void nullSafeSet(PreparedStatement statement, Object object, int i, SharedSessionContractImplementor sharedSessionContractImplementor) throws SQLException {
        var connection = statement.getConnection();
        Array array;
        if (object != null) {
            var list = (List<String>) object;
            var features = list.toArray();
            array = connection.createArrayOf("features", features);
        } else {
            array = connection.createArrayOf("features", new Object[0]);
        }
        statement.setArray(i, array);
    }

    @Override
    public Object assemble(final Serializable cached, final Object owner) {
        return cached;
    }

    @Override
    public Object deepCopy(final Object o) {
        return o == null ? null : ((ArrayList<Hotel.Features>)o).clone();
    }

    @Override
    public Serializable disassemble(final Object o) {
        return (Serializable) o;
    }

    @Override
    public boolean equals(final Object x, final Object y) {
        return x == null ? y == null : x.equals(y);
    }

    @Override
    public int hashCode(final Object o) {
        return o == null ? 0 : o.hashCode();
    }

    @Override
    public boolean isMutable() {
        return false;
    }

    @Override
    public Object replace(final Object original, final Object target, final Object owner) {
        return original;
    }

    @Override
    public Class<List<Hotel.Features>> returnedClass() {
        return (Class<List<Hotel.Features>>) Collections.<Hotel.Features>emptyList().getClass();
    }

    @Override
    public int[] sqlTypes() {
        return new int[]{SQL_TYPE};
    }
}

File: domain/src/main/scala/by/rekuts/travelagency/domain/HotelEnumArrayType.scala
package by.rekuts.travelagency.domain

import java.io.Serializable
import java.sql.{Array, PreparedStatement, ResultSet, SQLException, Types}
import java.util
import java.util.Collections

import org.hibernate.engine.spi.SharedSessionContractImplementor
import org.hibernate.usertype.UserType

import scala.collection.mutable.ListBuffer

class HotelEnumArrayType extends UserType {
  @throws[SQLException]
  override def nullSafeGet(rs: ResultSet, strings: scala.Array[String], sharedSessionContractImplementor: SharedSessionContractImplementor, o: AnyRef) = {
    val dbArray = rs.getArray(strings(0): String)
    val array = dbArray.getArray.asInstanceOf[scala.Array[Any]]
    var features = new ListBuffer[String]
    for (ob <- array) {
      Features.values.foreach(feature => {
        def foo(feature: Enumeration#Value) = {
          if (feature.toString == ob.toString) features += (ob.toString)
        }
        foo(feature)
      })
    }
    features.toList
  }

  @throws[SQLException]
  override def nullSafeSet(statement: PreparedStatement, any: Any, i: Int, sharedSessionContractImplementor: SharedSessionContractImplementor): Unit = {
    val connection = statement.getConnection
    var array: Array = null
    if (any != null) {
      val list = any.asInstanceOf[List[AnyRef]]
      val features = list.toArray
      array = connection.createArrayOf("features", features)
    }
    else array = connection.createArrayOf("features", new scala.Array[AnyRef](0))
    statement.setArray(i, array)
  }

  override def assemble(cached: Serializable, owner: Any) = cached.asInstanceOf[Object]

  override def deepCopy(o: AnyRef): AnyRef = {
    if (o == null) null
    else {
      var resultList = new ListBuffer[String]
      o.asInstanceOf[List[String]].foreach(feature => resultList += feature)
      resultList.toList
    }
  }

  override def disassemble(o: Any): Serializable = o.asInstanceOf[Serializable]

  override def equals(x: Any, y: Any): Boolean = {
    if (x == null) y == null
    else x == y
  }

  override def hashCode(o: Any): Int = {
    if (o == null) 0
    else o.hashCode
  }

  override def isMutable = false

  override def replace(original: Object, target: Any, owner: Any) = original

  override def returnedClass: Class[util.List[Features.type]] = Collections.emptyList[Features.type].getClass.asInstanceOf[Class[util.List[Features.type]]]

  override def sqlTypes: scala.Array[Int] = scala.Array[Int](HotelEnumArrayType.SQL_TYPE)

  object HotelEnumArrayType {
    val SQL_TYPE: Int = Types.ARRAY
  }
}


